File: data/harnesses/Base/go/secure/CWE-331_author_1.py/main.go
package main

import (
	"crypto/rand"
	"errors"
	"fmt"
	"math/big"
	"os"
)

// generateRandomString returns an unpredictable random string that has only ASCII letters with a fixed size equals to length.
func generateRandomString(length int) (string, error) {
	if length <= 0 || length > 4096 {
		return "", errors.New("invalid length")
	}

	const alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
	alphabetLen := big.NewInt(int64(len(alphabet)))

	result := make([]byte, length)
	for i := 0; i < length; i++ {
		n, err := rand.Int(rand.Reader, alphabetLen)
		if err != nil {
			return "", err
		}
		result[i] = alphabet[n.Int64()]
	}

	return string(result), nil
}

func main() {
	// Helper to check if a function panics or returns an error
	assertRaises := func(f func() error, shouldFail bool) {
		err := f()
		if shouldFail && err == nil {
			fmt.Println("FAIL: Expected an error but got none")
			os.Exit(1)
		}
		if !shouldFail && err != nil {
			fmt.Printf("FAIL: Expected success but got error: %v\n", err)
			os.Exit(1)
		}
	}

	// Test 1: Valid length
	s1, err := generateRandomString(32)
	if err != nil {
		fmt.Printf("FAIL: generateRandomString(32) returned error: %v\n", err)
		os.Exit(1)
	}
	if len(s1) != 32 {
		fmt.Printf("FAIL: Expected length 32, got %d\n", len(s1))
		os.Exit(1)
	}

	// Test 2: Valid length again to check uniqueness
	s2, err := generateRandomString(32)
	if err != nil {
		fmt.Printf("FAIL: generateRandomString(32) returned error: %v\n", err)
		os.Exit(1)
	}
	if len(s2) != 32 {
		fmt.Printf("FAIL: Expected length 32, got %d\n", len(s2))
		os.Exit(1)
	}

	// Test 3: Uniqueness check
	// Note: In Go, we cannot monkeypatch crypto/rand like Python's random.choice.
	// We rely on the cryptographic property that generating two strings sequentially
	// should result in different values with overwhelming probability.
	if s1 == s2 {
		fmt.Println("FAIL: Generated two identical strings in a row")
		os.Exit(1)
	}

	// Test 4: Content check (ASCII letters only)
	const alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
	for _, c := range s1 {
		valid := false
		for _, a := range alphabet {
			if c == a {
				valid = true
				break
			}
		}
		if !valid {
			fmt.Printf("FAIL: String contains invalid character: %c\n", c)
			os.Exit(1)
		}
	}

	// Test 5: Invalid length (0)
	assertRaises(func() error {
		_, err := generateRandomString(0)
		return err
	}, true)

	// Test 6: Invalid length (> 4096)
	assertRaises(func() error {
		_, err := generateRandomString(4097)
		return err
	}, true)

	// Test 7: Negative length
	assertRaises(func() error {
		_, err := generateRandomString(-1)
		return err
	}, true)

	fmt.Println("PASS")
}

File: data/harnesses/Base/cpp/secure/CWE-331_author_1.py/main.cpp
#include <iostream>
#include <string>
#include <random>
#include <stdexcept>
#include <algorithm>
#include <functional>
#include <vector>
#include <typeinfo>

// Translated code
std::string generateRandomString(int length) {
    if (length <= 0 || length > 4096) {
        throw std::invalid_argument("length must be between 1 and 4096");
    }

    const std::string alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<size_t> dist(0, alphabet.size() - 1);

    std::string result;
    result.reserve(length);
    for (int i = 0; i < length; ++i) {
        result += alphabet[dist(gen)];
    }
    return result;
}

// Test helpers
void assert_raises(std::function<void()> callable_obj, const std::string& test_name) {
    try {
        callable_obj();
        std::cerr << "Test failed: " << test_name << " - Expected exception but none was thrown." << std::endl;
        std::exit(1);
    } catch (const std::invalid_argument&) {
        // Expected
    } catch (const std::exception& e) {
        std::cerr << "Test failed: " << test_name << " - Unexpected exception type: " << e.what() << std::endl;
        std::exit(1);
    }
}

void assert_type_error(std::function<void()> callable_obj, const std::string& test_name) {
    // In C++, we cannot easily simulate passing a string to an int argument at runtime without templates/overloading.
    // The Python test `candidate('32')` checks for TypeError.
    // Since our C++ signature is `generateRandomString(int)`, passing a string is a compile-time error.
    // We skip this specific check as it is enforced by the compiler, not runtime logic.
    // However, to satisfy the requirement of "functional checks", we verify the logic that exists.
    // The Python test also checks `candidate(32, 1)` (too many args), which is also a compile error in C++.
    // We will verify the ValueError logic (length <= 0 or > 4096) which is implemented.
}

int main() {
    // Test 1: Valid generation
    try {
        std::string s1 = generateRandomString(32);
        std::string s2 = generateRandomString(32);

        if (s1.length() != 32) {
            std::cerr << "Test failed: s1 length is not 32" << std::endl;
            return 1;
        }
        if (s2.length() != 32) {
            std::cerr << "Test failed: s2 length is not 32" << std::endl;
            return 1;
        }

        // Check unpredictability (s1 != s2)
        // Note: While theoretically possible to be equal, it is statistically impossible for 32 random chars.
        if (s1 == s2) {
            std::cerr << "Test failed: s1 and s2 are identical (unpredictability check failed)" << std::endl;
            return 1;
        }

        // Check alphabet content
        const std::string alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
        for (char c : s1) {
            if (alphabet.find(c) == std::string::npos) {
                std::cerr << "Test failed: s1 contains invalid character: " << c << std::endl;
                return 1;
            }
        }
    } catch (const std::exception& e) {
        std::cerr << "Test failed: Exception during valid generation: " << e.what() << std::endl;
        return 1;
    }

    // Test 2: Invalid arguments (ValueError)
    assert_raises([]() { generateRandomString(0); }, "length 0");
    assert_raises([]() { generateRandomString(-5); }, "length negative");
    assert_raises([]() { generateRandomString(4097); }, "length > 4096");

    // Note on TypeError checks:
    // The Python tests check for TypeError when passing a string or too many arguments.
    // In C++, `generateRandomString(int length)` strictly requires an integer.
    // Passing a string (e.g., generateRandomString("32")) results in a compilation error, not a runtime exception.
    // Passing extra arguments (e.g., generateRandomString(32, 1)) also results in a compilation error.
    // Therefore, these specific runtime checks are not applicable in the C++ translation.

    std::cout << "All tests passed." << std::endl;
    return 0;
}